crates/ra_assists/src/handlers/fill_match_arms.rs

```diff
@@ -56,31 +50,115 @@ pub(crate) fn fill_match_arms(ctx: AssistCtx) -> Option<Assist> {
+        let mut arms: Vec<MatchArm> = match_arm_list.arms().collect();
+        if arms.len() == 1 {
+            if let Some(Pat::PlaceholderPat(..)) = arms[0].pat() {
+                arms.clear();
+            }
+        }
 
-        let new_arm_list = {
-            let arms = variants
-                .into_iter()
-                .filter_map(|variant| build_pat(db, module, variant))
-                .map(|pat| make::match_arm(iter::once(pat), make::expr_unit()));
-            indent_level.increase_indent(make::match_arm_list(arms))
-        };
+        let mut has_partial_match = false;
+        let variants: Vec<MatchArm> = variants
+            .into_iter()
+            .filter_map(|variant| build_pat(db, module, variant))
+            .filter(|variant_pat| {
+                !arms.iter().filter_map(|arm| arm.pat().map(|_| arm)).any(|arm| {
+                    let pat = arm.pat().unwrap();
+
+                    // Special casee OrPat as separate top-level pats
+                    let pats: Vec<Pat> = match Pat::from(pat.clone()) {
+                        Pat::OrPat(pats) => pats.pats().collect::<Vec<_>>(),
+                        _ => vec![pat],
+                    };
+
+                    pats.iter().any(|pat| {
+                        match does_arm_pat_match_variant(pat, arm.guard(), variant_pat) {
+                            ArmMatch::Yes => true,
+                            ArmMatch::No => false,
+                            ArmMatch::Partial => {
+                                has_partial_match = true;
+                                true
+                            }
+                        }
+                    })
+                })
+            })
+            .map(|pat| make::match_arm(iter::once(pat), make::expr_unit()))
+            .collect();
+
+        arms.extend(variants);
+        if has_partial_match {
+            arms.push(make::match_arm(
+                iter::once(make::placeholder_pat().into()),
+                make::expr_unit(),
+            ));
+        }
+
+        let indent_level = IndentLevel::from_node(match_arm_list.syntax());
+        let new_arm_list = indent_level.increase_indent(make::match_arm_list(arms));
 
         edit.target(match_expr.syntax().text_range());
         edit.set_cursor(expr.syntax().text_range().start());
         edit.replace_ast(match_arm_list, new_arm_list);
     })
 }
 
-fn is_trivial(arm: &ast::MatchArm) -> bool {
-    match arm.pat() {
-        Some(ast::Pat::PlaceholderPat(..)) => true,
-        _ => false,
+enum ArmMatch {
+    Yes,
+    No,
+    Partial,
+}
+
+fn does_arm_pat_match_variant(arm: &Pat, arm_guard: Option<MatchGuard>, var: &Pat) -> ArmMatch {
+    let arm = flatten_pats(arm.clone());
+    let var = flatten_pats(var.clone());
+    let mut arm = arm.iter();
+    let mut var = var.iter();
+
+    // If the first part of the Pat don't match, there's no match
+    match (arm.next(), var.next()) {
```

Comment:
Interesting, that this assist does not work if I comment out a required arm here.
Same for the next two matches (including the boolean one)

Reply:
This case doesn't work (the next one doesn't either) because it only works on Enums. We use the `enum_def` to know all the variants we have to suggest.

I'm not sure if there's a good way to be able to generically suggest this for other types without special casing.

Reply:
Ah, thanks for the elaboration, this makes sense now indeed.